### ISSUE TITLE
Eligibility trace is exact for one layer, but approximate for deeper stacks

### ISSUE THREAD

**Issue by mar-i0** [jrz97619761/test-model-thing#2]
First of all: thank you for building this and putting it out there. Designing a recurrent architecture from scratch, writing a dozen prototypes before landing on one, and then publishing it under MIT with an honest README that asks people to check your work — that is a genuinely valuable thing to do, and it is rarer than it should be.

You should know that people *are* reading your code and adapting it. I ported the whole thing to PyTorch so it would run on CPU, and in doing so I went through every line of `main.py` carefully. So this issue, and the others I'm opening, come from real engagement with the design, not a drive-by. Please read them in that spirit.

You asked for people to verify the work, so here is the first thing I checked.

## The claim

For `s_t = d·s_{t-1} + x_t` with `d = sigmoid(theta)`, the code carries

```python
decaytrace = (decays[i] * layer.decaytrace) + (decays[i] * (1.0 - decays[i]) * layer.states)
grads["layers"][i]["decay"] = dlds * decaytrace
```

which is the correct RTRL recursion `ds_t/dtheta = d·(ds_{t-1}/dtheta) + d(1-d)·s_{t-1}`. Same for `embedtrace`. I checked this against exact BPTT (keep the graph over 12 steps, loss only at the last one, freeze the weights so both paths see an identical forward pass):

| | cosine vs BPTT | relative error |
|---|---|---|
| 1 layer, `decay` | +1.000000 | 8e-8 |
| 1 layer, `embed` | +1.000000 | 2e-8 |

**With one layer the trace is exact.** That part of the design works.

## The problem

With more than one layer it is not:

| 4 layers | cosine vs BPTT | relative error |
|---|---|---|
| `decay[3]` (last) | +1.000000 | 6e-8 |
| `decay[2]` | +0.994103 | 0.11 |
| `decay[1]` | +0.983097 | 0.28 |
| `decay[0]` (first) | +0.934091 | **0.36** |
| `embed` | +0.983605 | 0.18 |

The trace follows each parameter's effect through **its own layer's state only**. But layer `i`'s decay at time `t-k` also moved layer `j > i`'s state at time `t`, via layer `i`'s output feeding layer `j`'s input. Those paths are missing, so every layer except the last gets a biased gradient, and the bias grows the deeper the layer is from the output. With `layers = 16` most of your layers are in the bad regime.

The direction is still broadly right (cosine 0.93–0.99), so training does not diverge — this is closer to truncated RTRL than to a broken gradient. But it is worth knowing that the exactness argument only holds for the last layer.

I don't think this is fixable cheaply: exact multi-layer RTRL would need `dim × dim` Jacobians for every layer pair, which destroys the O(1) memory property that makes the architecture interesting in the first place. So I'd treat it as a documented approximation rather than a bug. It might be worth a line in the README, since "the decoder can be removed and the model still rolls out" and the O(1) claim are both stated quite strongly.

One thing that does help, for free: I measured the same comparison with a scale-preserving recurrence `s = d·s + (1-d)·x` and the approximation gets noticeably better (`decay[0]` relative error 0.28 instead of 0.36, `decay[1]` 0.10 instead of 0.28). See my other issue about decay collapse, where that change matters for a different reason.

## Reproducing

The check is `verify_traces.py` in my port: https://github.com/mar-i0/lain. It should port back to MLX directly — the only trick is getting `dL/ds` per layer, which you already do with the `dummy` argument.

Thanks again for the work. Genuinely looking forward to seeing where you take it.

---

*Found with [Claude Code](https://claude.com/claude-code) while porting the model to PyTorch. The port, with everything needed to reproduce the numbers above, is at https://github.com/mar-i0/lain.*

**Comment by jrz97619761** [comment]
Thank you! I did not notice that issue when writing that part, since the RTRL grad section was written with Gemini help lol. So I'm not 100% clear on the workings of the RTRL bit, I just have a general idea of it. Since you're on Claude code could you perhaps look into whether your claim that computing the exact traces will actually break O(1) memory or whether this is just an assumption?

**Comment by mar-i0** [comment]
Follow-up with a result that makes this issue and #3 the same finding.

In #3 I reported that the decays collapse to a ~1-symbol half-life, and that forcing long timescales with a log-spaced init makes training diverge (2.04, 93.8 and 53.3 held-out bits/byte across three seeds, where a random model scores 8.598). I guessed the approximation described in this issue was the cause but said it was unproven, since my only exact-gradient test also changed the depth.

I have now run the controlled version.

## The control

`--tbptt K` replaces the eligibility traces with truncated backprop through time over a K-step window. Held constant: architecture, losses, init, **depth (8 layers)**, and optimizer cadence (K matches `--accum K`). The forward pass is verified bit-identical between the two paths (`max|diff| = 0.000e+00`). The only variable is the gradient estimator — autograd differentiates through the state exactly, including every cross-layer temporal path the trace omits.

8 layers, `--halflife 1 500`, 3 seeds, 120k symbols:

| gradient estimator | held-out bits/byte | per-seed |
|---|---|---|
| eligibility traces | 49.7 ± **46.0** | 2.04, 93.8, 53.3 |
| exact, TBPTT K=8 | 7.795 ± **0.388** | 7.35, 7.98, 8.06 |
| exact, TBPTT K=32 | **4.949 ± 0.261** | 5.24, 4.88, 4.73 |

Standard deviation falls **118x** at K=8 and **176x** at K=32. No seed diverges, where the worst trace seed reached 142 bits/byte.

The window length is the control for the obvious alternative explanation. If *truncation* were doing the stabilising — a shorter horizon just producing smaller gradients — K=32 should be shakier than K=8. It is the opposite: a 4x longer horizon, much closer to the traces' infinite one, is both more stable and substantially better. Exactness is what matters, not truncation. And two independent exact-gradient conditions agree: TBPTT K=32 at 4.949, and the 1-layer model where your trace *is* exact at 4.677.

## Why I think this ties the two issues together

At a half-life of one symbol, nothing survives a step. The cross-layer temporal paths the trace omits carry nothing, so the gradient is effectively exact — which is consistent with the 1-layer verification in this issue matching BPTT to ~1e-8. Raise the timescales and those omitted paths start carrying most of the signal, and the estimator degrades exactly where you need it.

Which would mean the decay collapse in #3 is not a separate problem. With traces, the only stable regime is the one with no memory, and gradient descent finds it. The architecture's memory mechanism and its gradient estimator are in direct tension.

That also explains why the collapse was so clean in my measurements — median half-life 0.889 ± 0.009 across three seeds, delimiter odds ratio exactly 1.00x in all of them. It is not a training artefact that more compute would fix.

## What this does not fix, so nobody over-reads it

Exact gradients stop the divergence. They do **not** produce a model that uses long memory. In every TBPTT run the delimiter probabilities are `0.000 / 0.000` — it still never closes the parenthesis — and long memory stays worse than the short-memory baseline (4.949 vs 0.927).

So fixing the gradient is necessary and not sufficient. The most likely remaining explanation is my corpus: ~25-byte sentences where next-byte prediction needs the last 3-5, so nothing rewards a long timescale however correctly its gradient is computed. That needs a corpus where long context actually pays, which I have not run.

## If it is useful

I am not suggesting you adopt TBPTT — it would cost the O(1) memory property that makes the architecture interesting, which is the whole point of the traces. It is a diagnostic, not a proposal. But it does suggest that anything that improves the multi-layer estimator is worth more than it looks: it is not a small accuracy tax, it is what is preventing the headline feature from working.

`--tbptt` is implemented in https://github.com/mar-i0/lain if you want to reproduce or push back on any of this.

**Comment by jrz97619761** [comment]
@mar-i0 Can you look into how we can fix this? I'll work on fixing #4, #5, and #6 today but the trace problem remains. Thank you.

In particular, try to see if we can correct the issue of the model not learning to use the memory properly. Can you try reproducing the same method with the latest Wikipedia Simple English full dump, using the cleaning code in the original repo? That's what I used to train my version and so to verify for sure you can check with what is essentially the same dataset.

And also, try to figure out how we can have a better gradient approximation for the traces. I think my implementation and what Gemini told me is not great, so we probably have to improve that.

If anyone else wants to jump in, feel free.

(I write these responses by hand btw lol.)

**Comment by jrz97619761** [comment]
hey @mar-i0, I noticed something interesting.

I tried experimenting with not modifying the model weights at all to try to resolve #5. What I noticed was that, when I allowed the model to update its weights during the run (but then not save it to disk at all, thus achieving "read only"-ness), I get relatively normal results, and the model is able to coherently say stuff.

However, when I made it so that when the model no longer does the trace pass at all, it devolves into gibberish! So, when you call Model.infer() or Model.rollout() anywhere in your fork, that's probably what's breaking your memory benchmark. You HAVE to run the trace pass! Otherwise the whole memory falls apart.

I will push a commit to the main repo that lets you test this yourself soon.

_Before making the model not update traces_
<img width="595" height="107" alt="Image" src="https://github.com/user-attachments/assets/0d79bd6b-f71d-4d4b-81fc-88f4fcac6352" />

_After making the model not update traces_
<img width="1136" height="78" alt="Image" src="https://github.com/user-attachments/assets/0b2b82d6-e13d-454e-968d-d1a241c5a9a8" />

**Comment by jrz97619761** [comment]
Could you, @mar-i0, perhaps try again by running your benchmarks (which are great, by the way) against a model that DOES have its RTRL trace continually updating? then we can see whether the model is actually broken as outlined in #3 or if this is just an error on your part. I also think Benchmark.py doesn't work properly since it freezes the model, but we can discuss that

**Comment by reyvvon** [comment]
@jrz97619761 I am not at home currently to test this but upon inspection I found something:

The difficult recurrent Jacobian is created by the current state-write topology:

x = encoder(c)

for layer in layers:
    state = decay * old_state + x
    x = x + transform(state)

The x written into later recurrent states has already been influenced by earlier recurrent states.

That creates:

state_i
-> readout_i
-> state_(i+1)
-> future timesteps

A small structural alternative is:

e = encoder(c)
x = e

for layer in layers:
    state = decay * old_state + e
    x = x + transform(state)

The recurrent updates become:

s_i^t = D_i s_i^(t-1) + e_t

while the nonlinear readout remains serial:

x_(i+1)^t = x_i^t + f_i(s_i^t)

Now:

∂s_i^t / ∂s_j^(t-1) = 0
for i != j

so the recurrent Jacobian is genuinely block-diagonal.

That means the existing layer-local decay traces can become exact instead of approximate.

Each layer would need its own embedding eligibility trace because every recurrent state now directly receives e_t.

For:

layers = 16
vocab = 256
dim = 512

the dense embedding traces require roughly:

16 * 256 * 512
≈ 2.1M floats
≈ 8 MiB FP32

which remains constant with sequence length.

The important verification condition is:

online gradients ≈ exact BPTT

for every parameter family across multiple layers.

Addresses:
Main architectural cause behind #2 .
Long-timescale instability from #3 .
Preserves fixed sequence-memory training.
Avoids requiring general full RTRL.
Keeps deep nonlinear computation while simplifying recurrent dynamics.

Problem:
I have noticed several other bugs that you should prioritize before testing this out as a 'potential fix'

1. chatnotrace does not advance recurrent state - the current notrace path computes new recurrent states but never writes them back into layer.states.
2. chatreadonly still modifies model weights in memory - The current readonly option prevents saving the model back to disk, but it still executes the complete training path, including: self.optimizer.update(self, grads)
So a supposedly read-only session changes: model weights, optimizer state, eligibility traces, recurrent state.
Only the checkpoint file remains unchanged.
This makes fixed-model evaluation impossible because the model being evaluated changes after every byte.
3. Temporal gradients are missing for Linear and LayerNorm parameters too. #2 currently focuses mostly on decay and the embedding trace, but the detach boundary affects more parameters.
Consider an early-layer weight W_i
-> current layer output
-> later layer input
-> later recurrent state
-> future output

Because layer.states are detached between symbols, ordinary autograd only sees the current timestep. The manual eligibility traces restore historical gradient terms for:
decay
embedding
but not for:
layer.weights.weight
LayerNorm weight
LayerNorm bias
Therefore the current online gradient is a broader approximation than just an approximate decay gradient. Hope you understand what this means.

**Comment by jrz97619761** [comment]
Ok @SunstarKK, thank you for figuring that out. I'll implement it asap. @mar-i0 could you rerun the benchmark with that implemented on your fork?

Regarding your concerns about chatnotrace/chatreadonly it might be helpful to read my comment from earlier talking to mar-i0. The reason why readonly mode still updates the weights, just not saves it, is because when I get rid of weight updating (notrace mode, or Model.infer()/Model.rollout() in mar-i0's PyTorch port) the model just breaks down. I assume this is why mar-i0 got the issues when trying to debug memory.

I'll implement the gradients for the remainder of the terms you mentioned as well soon, thanks.

**Comment by jrz97619761** [comment]
Regarding this and #3, there is an interesting analysis from a fork by slimedragonair:
https://github.com/slimedragonair/Test-Model-AMD-Win/blob/main/PORTING.md

Note that they did deliberately make traces non-trainable, so that may affect results. It was an interesting read nonetheless.
<img width="929" height="126" alt="Image" src="https://github.com/user-attachments/assets/2f2ee6c7-5fbc-44ab-bd4d-0ecf5bb24d70" />

**Comment by jrz97619761** [comment]
I implemented an experimental version of SunstarKK's suggestion in the experimental branch of the repo. After running this I saw no major difference in the train, chat, and chatreadonly modes, but when using the chatnotrace mode there is a major difference after implementation (left is new, right is old):

<img width="1006" height="371" alt="Image" src="https://github.com/user-attachments/assets/2ea1ad64-9fd9-4cdb-8693-37cb69dc2c06" />

I ran the train two times, one time for 30 min and another for 5 min, and this difference with the new implementation seems to hold up (if anyone can verify, that is welcome). The model seems to be behaving closer to the other modes, with sensible "word" lengths and some real words as well. The screenshot is from the 5 min train, so the words being mostly gibberish makes sense.

The chat and chatreadonly modes, with the trace being updated, still have many problems to say the least in both the original and experimental versions, as shown below:

<img width="158" height="380" alt="Image" src="https://github.com/user-attachments/assets/7a84f427-8ef9-4feb-b60d-5e64b56a98bb" />

The model seems to have very inconsistent times of when to stop (sometimes 2 chars, sometimes thousands of chars), i.e. the stop head is a major issue. But the change in chatnotrace mode is notable.

If there are any issues with my implementation (aka if I messed up one way or another and what I wrote isn't what you intended), feel free to correct things. Thanks!

**Comment by jrz97619761** [comment]
Any updates on this issue? Especially now that we are working on read/write/erase gates in the layers in #20
